include/boost/url/decoded_view.hpp

```diff
@@ -0,0 +1,648 @@
+
+        @par Exception Safety
+        Strong guarantee.
+        Exceptions thrown on invalid positions.
+
+        @throw std::out_of_range If `pos > size()`
+
+        @param dest pointer to the destination character string
+        @param count requested substring length
+        @param pos position of the first character
+
+        @return Number of characters copied
+    */
+    BOOST_URL_DECL
+    size_type
+    copy(char* dest,
+         size_type count,
+         size_type pos = 0) const;
+
+    /** Return true if the corresponding comparison holds
+
+        The length of the sequences to compare is the smaller of
+        `size()` and `other.size()`.
+
+        The function compares the two strings as if by calling
+        `char_traits<char>::compare(to_string().data(), v.data(), rlen)`.
+
+        @param other string to compare
+
+        @return Negative value if this string is less than the other
+        character sequence, zero if the both character sequences are
+        equal, positive value if this string is greater than the other
+        character sequence
+    */
+    BOOST_URL_DECL
+    int
+    compare(string_view other) const noexcept;
+
+    /// @copydoc compare()
+    int
+    compare(char const* other) const noexcept
+    {
+        return compare(string_view(other));
+    }
+
+    /** Return true if the corresponding comparison holds
+
+        The length of the sequences to compare is the smaller of
+        `size()` and `other.size()`.
+
+        The function compares the two strings as if by calling
+        `char_traits<char>::compare(to_string().data(), v.to_string().data(), rlen)`.
+
+        @param other string to compare
+
+        @return Negative value if this string is less than the other
+        character sequence, zero if the both character sequences are
+        equal, positive value if this string is greater than the other
+        character sequence
+    */
+    BOOST_URL_DECL
+    int
+    compare(decoded_view other) const noexcept;
+
+    /** Returns a string with the decoded data
+    */
+    std::string
+    to_string() const
+    {
+        std::string r;
+        assign_to(r);
+        return r;
+    }
+
+    /** Assigns the decoded characters to a container
+
+        This overload allows the caller to recycle
+        the input string.
+
+        This function only participates in
+        overload resolution if the container
+        type `T` has the member function
+        `assign(iterator, iterator)`.
+
+        @return A string representing the
+        entire contents of the decoded range.
+    */
+    template <class String>
+    String&
+    assign_to(String& s) const;
+
+    /** Appends the decoded characters to a container
+
+        This overload allows the caller to recycle
+        the input string.
+
+        This function only participates in
+        overload resolution if the container
+        type `T` has the member function
+        `append(iterator, iterator)`.
```

Comment:
Same here. Don't use SFINAE use BOOST_STATIC_ASSERT. Your javadoc says `T` when you mean `String`. You should require `String` have a nested `value_type` that is `char`.

Reply:
lets rename this to `MutableString` and add a trait `is_mutable_string` which we can use in the static assert

Reply:
We already have `detail::has_assign_from` and `detail::has_append_from` to make the static assert easier to read. Should `is_mutable_string` include both conditions? Should we make it public? Where should it go?